src/nvim/msgpack_rpc/channel.c

```diff
@@ -313,27 +312,39 @@ static void handle_request(Channel *channel, msgpack_object *request)
     return;
   }
 
-  // Retrieve the request handler
   MsgpackRpcRequestHandler handler;
-  Array args = ARRAY_DICT_INIT;
   msgpack_object *method = msgpack_rpc_method(request);
+  handler = msgpack_rpc_get_handler_for(method->via.bin.ptr,
+                                        method->via.bin.size);
 
-  if (method) {
-    handler = msgpack_rpc_get_handler_for(method->via.bin.ptr,
-                                          method->via.bin.size);
-    if (handler.fn == msgpack_rpc_handle_missing_method) {
-      String m = method->via.bin.size > 0
-        ? cbuf_to_string(method->via.bin.ptr, method->via.bin.size)
-        : cstr_to_string("<empty>");
-      ADD(args, STRING_OBJ(m));
-      handler.async = true;
-    } else if (!msgpack_rpc_to_array(msgpack_rpc_args(request), &args)) {
-      handler.fn = msgpack_rpc_handle_invalid_arguments;
-      handler.async = true;
-    }
-  } else {
-    handler.fn = msgpack_rpc_handle_missing_method;
-    handler.async = true;
+  // check method arguments
+  Array args = ARRAY_DICT_INIT;
+  if (!msgpack_rpc_to_array(msgpack_rpc_args(request), &args)
+      && !ERROR_SET(&error)) {
+    api_set_error(&error, kErrorTypeException, "Invalid method arguments");
+  }
+
+  // check method
+  if (handler.fn == msgpack_rpc_handle_missing_method && !ERROR_SET(&error)) {
+    size_t method_size = method->via.bin.size;
+    String format_string = method_size > 0 ?
+      cbuf_to_string(method->via.bin.ptr, method->via.bin.size) :
+      cstr_to_string("<empty>");
+    api_set_error(&error, kErrorTypeException, "Invalid method: %s",
+                  format_string);
+    xfree(format_string.data);
+  }
+
+  if (ERROR_SET(&error)) {
+    channel_write(channel,
```

Comment:
`send_error(channel, request_id, error.msg)` could be used here.  Object `error` could be replaced with a char pointer for the error message.

Reply:
Ah, ok! Should one also use it in line 299 then?

Reply:
`send_error()` needs to be modified for this to return the result of `channel_write()`.

Reply:
Ah i see, maybe thats not the scope of this refactoring then!